File: Assignment5/src/main/java/BankAccount.java
import java.util.ArrayList;
import java.util.List;

public class BankAccount {

    private final String owner;
    private double balance;
    private final List<String> transactions = new ArrayList<>();

    public BankAccount(String owner, double initialBalance) {
        if (initialBalance < 0) throw new IllegalArgumentException("Initial balance cannot be negative");
        this.owner = owner;
        this.balance = initialBalance;
        if (initialBalance > 0) transactions.add("Initial deposit: +" + initialBalance);
    }

    public BankAccount(String owner) {
        this(owner, 0);
    }

    public double deposit(double amount) {
        if (amount <= 0) throw new IllegalArgumentException("Deposit amount must be positive");
        balance += amount;
        transactions.add("Deposit: +" + amount);
        return balance;
    }

    public double withdraw(double amount) {
        if (amount <= 0) throw new IllegalArgumentException("Withdrawal amount must be positive");
        if (amount > balance) throw new IllegalArgumentException("Insufficient funds");
        balance -= amount;
        transactions.add("Withdrawal: -" + amount);
        return balance;
    }

    public double getBalance() { return balance; }
    public String getOwner()   { return owner; }

    public List<String> getTransactionHistory() {
        return new ArrayList<>(transactions);
    }
}

File: Assignment5/BankAccount.py
class BankAccount:
    def __init__(self, owner, initial_balance=0):
        if initial_balance < 0:
            raise ValueError("Initial balance cannot be negative")
        self.owner = owner
        self._balance = initial_balance
        self._transactions = []
        if initial_balance > 0:
            self._transactions.append(f"Initial deposit: +{initial_balance}")

    def deposit(self, amount):
        if amount <= 0:
            raise ValueError("Deposit amount must be positive")
        self._balance += amount
        self._transactions.append(f"Deposit: +{amount}")
        return self._balance

    def withdraw(self, amount):
        if amount <= 0:
            raise ValueError("Withdrawal amount must be positive")
        if amount > self._balance:
            raise ValueError("Insufficient funds")
        self._balance -= amount
        self._transactions.append(f"Withdrawal: -{amount}")
        return self._balance

    def get_balance(self):
        return self._balance

    def get_history(self):
        return list(self._transactions)


if __name__ == "__main__":
    account = BankAccount("Daniil", 1000)
    account.deposit(500)
    account.withdraw(200)
    print(f"Owner:   {account.owner}")
    print(f"Balance: {account.get_balance()}")
    print("History:")
    for entry in account.get_history():
        print(f"  {entry}")
